mfsmanpages/mfschunkdbdump.8
.TH mfschunkdbdump "8" "May 2026" "MooseFS 4.59.2-1" "This is part of MooseFS"
.ss 12 0
.SH NAME
mfschunkdbdump \- dumps data stored by a chunkserver in a chunkdb file
.SH SYNOPSIS
.B mfschunkdbdump \fIchunkdb_filename\fP
.SH DESCRIPTION
.PP
\fBmfschunkdbdump\fP dumps data, stored by a chunkserver in a chunkdb file, into human-readable (txt) format.
.SH "REPORTING BUGS"
Report bugs to <bugs@moosefs.com>.
.SH COPYRIGHT
Copyright (C) 2026 Jakub Kruszona-Zawadzki, Saglabs SA

This file is part of MooseFS.

MooseFS is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 2 (only).

MooseFS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, see
<https://www.gnu.org/licenses/>.
.SH "SEE ALSO"
.BR mfschunkserver (8)


NEWS
This file lists noteworthy changes in MooseFS.

* MooseFS 4.59.2-1 (2026-05-14)

  - (cs) fixed bug preventing background chunk testing

* MooseFS 4.59.1-1 (2026-05-04)

  - (cs) fixed potential OOB read

* MooseFS 4.59.0-1 (2026-04-23)

  - (all) changed licence handling (pro only)
  - (all) fixed some security issues

* MooseFS 4.58.5-1 (2026-04-01)

  - (cs) removed aggresive fadvise calls

* MooseFS 4.58.4-1 (2026-03-18)

  - (all) added spec files for Fedora, changed autogen.sh to bootstrap.sh
  - (all) added 'printflike' macro to mfslog, fixed all printflike macros, fixed some syslog args
  - (common) silenced address sanitizer unnecessary warning
  - (common) fixed masterresolver (potential segfault when the last name was empty - pro only)
  - (common) added protection against double initialization of strerr
  - (master) added expected chunk versions to syslog messages

* MooseFS 4.58.3-1 (2025-12-15)

  - (master+tools) fixed conditions for reversible storage class
  - (daemons) fixed small security issue (clearing auxiliary groups before setuid)
  - (libmfsio) fixed descriptor resize function

* MooseFS 4.58.2-1 (2025-11-03)

  - (master) changed timeouts used in master to master connections (pro only)
  - (client) fixed test for overlapping requests in read-ahead algorithm (pull request by Chuandev)
  - (all) fixed typos and spelling errors

* MooseFS 4.58.1-1 (2025-09-29)

  - (master) fix for segfault in flocklocks.c
  - (cli) fixed handling disconnected chunkservers
  - (cli) import fixes in cli mode
  - (tools) improved error messages in mfssetfacl
  - (tools) fixed parsing numerical groups in mfssetfacl
  - (master) prevent setting archive flag for files that are already in trash

* MooseFS 4.58.0-1 (2025-09-15)

  - (master) fixed support for partial readdir
  - (client) added reading dirs in parts
  - (client) added listxattr cache
  - (client) redesigned readdir caches
  - (client) added 'mfsreaddirplusminto' option needed for readdirplus

* MooseFS 4.57.8-1 (2025-08-05)

  - (crc) fixed crc calc for CPU's that can't access memory on non even addresses (i.e. arm-hf)
  - (cs) less disk ops for idle servers
  - (debian) updated postinstall scripts
  - (nbd) fixed restoring nbd devices after sleep/hibernate
  - (cli) fixed displaying master's and chunkserver's charts data
  - (systemd) added start/stop timeouts to chunkserver systemd parameters

* MooseFS 4.57.7-1 (2025-06-16)

  - (all) joined packages cgi and cgiserv to a new gui package
  - (gui) added dedicated GUI server
  - (gui) gui refactor
  - (cli) cli refactor
  - (gui) added support for Prometheus metrics
  - (master) fixed freeing waiting locks from data structures
  - (daemons) added condition to skip socket routines when poll returned zero
  - (daemons) fixed disabling OOM killer
  - (client) removed using potentially dangerous sprintf function

* MooseFS 4.57.6-1 (2025-03-25)

  - (master) fixed predefined storage classes
  - (master) fixed csipmap initialization issue (no csmap after start/restart)
  - (master) fixed lock listing (case of posix_locks and flocks being in the same packet)
  - (master) fixed chunk recovery conditions during read and lookup
  - (master) changed description of default storage classes
  - (man) added manpage for mfsmount.cfg file
  - (man) added subsections to all man pages and reworded some of them

* MooseFS 4.57.5-1 (2025-02-26)

  - (master) added the number of localsplit operations to the total split counter
  - (master) fixed chunk copies and EC matrix counters when both copies and parts are present
  - (master) fixed sending changelog rotate packet to metalogger (CE only)

* MooseFS 4.57.4-1 (2025-02-17)

  - (chunktool) added recursive mode and option to throw away damaged chunks

* MooseFS 4.57.3-1 (2025-02-04)

  - (mfsio) added support for xattr and acl
  - (master) added protection against setting too long xattr
  - (master) fixed case when user changes only grace period in quota

* MooseFS 4.57.2-1 (2025-01-27)

  - (mount) added 'fake' system.posix_acl* xattrs to listxattr (previously not needed)
  - (gui) added cluster traffic charts, updated names
  - (cli) added checks for wrong integer values in options
  - (freebsd) changed port base name in freebsd from moosefs4 to moosefs
  - (master) added KEEP_LEADERSHIP option
  - (gui) charts legend update

* MooseFS 4.57.1-1 (2025-01-08)

  - (master) fixed bug in case of making space for ec parts marked for removal
  - (gui) fixed tmp maintenance time display
  - (gui) fixed graph summary bar display
  - (master) fixed changelog rotation issues
  - (mfsio) fixed mutex initialisation issue
  - (autoconf) updated configure.ac for new PKG_PROG_PKG_CONFIG macro
  - (daemons) added timestamp to info file
  - (master) added meta info to info file
  - (gui) fixed asyncio for python <3.7

* MooseFS 4.57.0-1 (2024-12-09)

  - (all) added storage class priority, description and export group
  - (master) added storage class group allowance to exports.cfg (deprecates mingoal/maxgoal)
  - (master) changed predefined storage classes in new MFS instances
  - (all) removed support for python 2.7
  - (master) always resolve master hostname on info command (PRO only)
  - (master) added to supervisor in 'simple mode' broadcasting LEADER IP to all FOLLOWERs (PRO only)
  - (master) added info for changelog.c
  - (daemons) warnings if a variable is defined more than once in a config file
  - (bdev) added test for posix compliant ioctl prototype
  - (freebsd) moved port files from sysutils to filesystems
  - (master) fixed chunkserver registration in DEPUTY mode (PRO only)
  - (all) added names to some constants used across the system
  - (all) added new charts data (total data I/O from all clients)
  - (metasearch) added printing paths for trash and sustained items
  - (master) fixed constants in metadata.c

* MooseFS 4.56.6-1 (2024-09-23)

  - (cgi+cli) fixed bugs in CLI mode introduced by new CGI features

* MooseFS 4.56.5-1 (2024-09-18)

mfschunkserver/mfschunkdbdump.c
/*
 * Copyright (C) 2026 Jakub Kruszona-Zawadzki, Saglabs SA
 * 
 * This file is part of MooseFS.
 * 
 * MooseFS is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 (only).
 * 
 * MooseFS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see
 * <https://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "datapack.h"

int chunkdb_dump(const char *fullname) {
	struct stat sb;
	int fd;
	uint8_t *chunkbuff;
	const uint8_t *rptr,*rptrmem,*endbuff;
	uint16_t pleng;
	uint64_t chunkid;
	uint32_t version;
	uint16_t blocks;
	uint16_t pathid;
	uint16_t hdrsize;
	uint32_t diskusage;
	uint8_t testedflag;
	uint8_t mode;
	uint8_t rsize;
	int res;

	fd = open(fullname,O_RDONLY);
	if (fd<0) {
		fprintf(stderr,"can't open chunkdb file\n");
		return -1;
	}
	if (fstat(fd,&sb)<0) {
		close(fd);
		fprintf(stderr,"can't stat chunkdb file\n");
		return -1;
	}
	chunkbuff = malloc(sb.st_size);
	if (chunkbuff==NULL) {
		fprintf(stderr,"can't allocate memory\n");
		close(fd);
		return -1;
	}
	if (read(fd,chunkbuff,sb.st_size)!=sb.st_size) {
		fprintf(stderr,"can't read chunkdb file\n");
		free(chunkbuff);
		close(fd);
		return -1;
	}
	close(fd);

	rptr = chunkbuff;
	endbuff = rptr+sb.st_size;

	if (memcmp(rptr,"MFS CHUNKDB",11)!=0 && rptr[11]!='1' && rptr[11]!='2' && rptr[11]!='3' && rptr[11]!='4') {
		fprintf(stderr,"wrong chunkdb header\n");
		free(chunkbuff);
		return -1;
	}
	mode = rptr[11]-'0';
	if (mode==1) {
		rsize = 16;
	} else if (mode==2) {
		rsize = 18;
	} else if (mode==3) {
		rsize = 19;
	} else {
		rsize = 23;
	}
	rptr+=12;

	pleng = get16bit(&rptr);
	if (rptr+pleng>endbuff) {
		fprintf(stderr,"wrong path in chunkdb file\n");
		free(chunkbuff);
		return -1;
	}
	printf("path: ");
	while (pleng>0) {
		putchar(*rptr);
		rptr++;
		pleng--;
	}
	printf("\n");
	rptrmem = rptr;
	chunkid = 0;
	version = 0;
	blocks = 0;
	hdrsize = 1024+4096;
	pathid = 0xFFFF;
	testedflag = 0;
	diskusage = 0;

	res = 0;

	while (rptr+rsize<=endbuff) {
		chunkid = get64bit(&rptr);
		version = get32bit(&rptr);
		blocks = get16bit(&rptr);
		if (blocks>1024 && blocks!=0xFFFF) {
			printf("chunk %016"PRIX64"_%08"PRIX32" - wrong block count (%u)\n",chunkid,version,blocks);
			res = -1;
		}
		if (mode>=2) {
			hdrsize = get16bit(&rptr);
		}
		pathid = get16bit(&rptr);
		if (pathid>255) {
			printf("chunk %016"PRIX64"_%08"PRIX32" - wrong pathid (%u)\n",chunkid,version,pathid);
			res = -1;
		}
		if (mode>=3) {
			testedflag = get8bit(&rptr);
		}
		if (mode>=4) {
			diskusage = get32bit(&rptr);
		}
		if (chunkid==0) {
			break;
		}
	}
	if (rptr!=endbuff || chunkid!=0 || version!=0 || blocks!=0 || pathid!=0 || testedflag!=0 || diskusage!=0) {
		printf("wrong chunkdb ending\n");
		res = -1;
	}

	rptr = rptrmem;

	while (rptr+rsize<=endbuff) {
		chunkid = get64bit(&rptr);
		version = get32bit(&rptr);
		blocks = get16bit(&rptr);
		if (mode>=2) {
			hdrsize = get16bit(&rptr);
		}
		pathid = get16bit(&rptr);
		if (mode>=3) {
			testedflag = get8bit(&rptr);
		}
		if (mode>=4) {
			diskusage = get32bit(&rptr);
		}
		if (chunkid==0) {
			break;
		}
		if (pathid<256) {
			printf("%02"PRIX16"/chunk_%016"PRIX64"_%08"PRIX32".mfs ; length: ",pathid,chunkid,version);
			if ((mode>=2 && hdrsize==0) || blocks==0xFFFF) {
				printf("-");
			} else {
				printf("%u",hdrsize+blocks*65536U);
			}
			printf(" ; blocks: ");
			if (blocks!=0xFFFF) {
				printf("%u",blocks);
			} else {
				printf("-");
			}
			if (mode>=2) {
				printf(" ; hdrsize: ");
				if (mode>=2 && hdrsize==0) {
					printf("-");
				} else {
					printf("%u",hdrsize);
				}
			}
			if (mode>=3) {
				printf(" ; tested: %u",testedflag);
			}
			if (mode>=4) {
				printf(" ; diskusage: ");
				if (diskusage>0) {
					printf("%u",diskusage);
				} else {
					printf("-");
				}
			}
			printf("\n");
		} else {
			printf("pathid: %04"PRIX16" ; chunkid: %016"PRIX64" ; version: %08"PRIX32" ; hdrsize: %u ; blocks: %u ; tested: %u ; diskusage: %u\n",pathid,chunkid,version,hdrsize,blocks,testedflag,diskusage);
		}
	}

	free(chunkbuff);
	return res;
}

int main(int argc,char *argv[]) {
	if (argc!=2) {
		fprintf(stderr,"usage: %s <chunkdb>\n",argv[0]);
		return 1;
	}
	if (chunkdb_dump(argv[1])<0) {
		fprintf(stderr,"file has errors\n");
		return 1;
	}
	return 0;
}
